### DISCUSSION TITLE
#4 4th Sprint: Class Diagrams and Project Setup

### DISCUSSION THREAD

**Discussion by AlexMi-Ha** [CUMGroup/AMOGUS#20] [Blog]
Hello there AMOGUS gang ඞ,

this week we took on the challenge to define our service infrastructure and the architecture of our backend. 🔥

With the basic architecture we mainly followed the ["Clean Architecture"](https://github.com/jasontaylordev/CleanArchitecture) example by Jason Taylor with a few modifications and split our Backend in three big parts. The Api Layer, the Core Layer and the Infrastructure Layer.
<div align="center">
<img height=500 src="https://user-images.githubusercontent.com/61502536/200187268-5d652c81-02f7-4e2e-9427-bb303c9ccaa9.png" alt="Architecture"></img>
</div>

Additionally, we thought about our Service Infrastructure and started work on a basic class diagram, showing our Backend.
<div align="center">
<img height=700 src="https://user-images.githubusercontent.com/61502536/200187507-0d2ed68e-dad2-4620-b3d3-b61c91ae3793.png" alt="class diagram"></img>
</div>

When designing this diagram, we tried to split our system in as many Services as needed, therefore applying the SOLID principle SRP (Single Responsibility Principle).

Another SOLID principle we used is DIP (Dependency Inversion Principle) in the abstraction of the Infrastructure Layer with the Core Layer. The Api Layer can only communicate with the Infrastructure Layer via the Core Layer interfaces through Dependency Injection.👀

<hr/>

Furthermore, we started work on the ER-Diagram for our Database. We are confident that we can finish it in this sprint.

Lastly, we've updated the sequence diagrams of the UCRS accordingly.
See [Login UCRS](https://github.com/CUMGroup/AMOGUS/blob/main/documents/requirements/Use-Case-Realization%20Specification/ucrs_login.md), [Register UCRS](https://github.com/CUMGroup/AMOGUS/blob/main/documents/requirements/Use-Case-Realization%20Specification/ucrs_register.md).

As an additional Task in our next sprint we plan to rework our View-Mockups with a wireframe tool to make them look 🌟prettier🌟

But thats it for this weeks blog. ඞ

Yours truly,
CUMGroup

**Comment by EhrlerL** [comment]
Hello CUMGroup!

I really like your ABBA-circle. It is a very good addition to your class diagram and makes it easier to quickly understand your project's structure. 
Your class diagram is also well structured and easy to read. To me it seems like you used the UML components correctly.
Over all, very good job!
I'm already looking forward to see your finished ER-diagram.

Cheers, capyclue. 𐃶

**Reply by AlexMi-Ha** [reply]
Hi Team Capyclue,

I'm glad the diagrams are easy to understand and can describe what we are planning to do!
We've put a lot of work in the architecture and structure in order to have a solid foundation for the rest of our program.
Thanks for your feedback!

Yours truly
CUMGroup

**Comment by MichaelMGraf** [comment]
Dear AMOGUS Team,

first of all we have to give you credit for the very nice diagrams you have made.
A consistent and well thought out architecture is one of the most important things in software development.
The service architecture you introduced is very detailed and carefully chosen.
Your encapsulation principle through interfaces and dependency injection seems to be very useful for extensions and testing purposes.

We are curious to see how your service structure will develop with implemetation and further expansion of your system.

Much love,
Team DD (Don't Let You Die) ❤️

**Reply by Annalytic-programming** [reply]
Hello Team DD 🍌,

thanks for your lovely comment. 
Due to our consistency we hope to achieve a better overview for each of our artefacts and documents. It is an honor to us if you like this attitude.
Furthermore we are proud to have a really good backend engineer in our team, who is an expert in software architecture and leads the team here. 

Yours truly 
CUMGroup
